doc/man/man1/synco.1
\"    This documentation is copyright 1994 David Huron.
.TH synco 1 "1994 Dec. 4"
.AT 3
.sp 2
.SH "NAME"
.in +2
.in +11
.ti -11
\fBsynco\fR  --  measure degree of metric syncopation for Humdrum input
.in -11
.in -2
.sp 1
.sp 1
.SH "SYNOPSIS"
.in +2
.in +7
.ti -7
\fBsynco\fR  [\fB-e\fR]  [\fIinputfile ...\fR]  [ > \fIoutputfile.syn\fR]
.in -7
.in -2
.sp 1
.sp 1
.SH "DESCRIPTION"
.in +2
The
.B "synco"
command characterizes the degree of metric syncopation evident at
successive moments in a passage.
It outputs a single Humdrum spine (\f(CR**synco\fR) containing numerical
values representing the instantaneous level of syncopation.
The
.B "synco"
command requires at least two spines of input data -- one of which must be
**metpos.
(The \f(CR**metpos\fR representation encodes the position in the metric
hierarchy for each data record in the input.)
The other input spine(s) must contain information that implicitly or explicitly
encodes the occurrence of note onsets.
Appropriate inputs to
.B "synco"
include the pitch-related representations listed below.
For descriptions of the various input representations
refer to Section 2
.I "(Representation Reference)"
of this reference manual.
.sp 1
.TS
l l.
\f(CR**cbr\fR	critical band rate (in equivalent rectangular
	  bandwidths)
\f(CR**cents\fR	hundredths of a semitone with respect to middle C=0
\f(CR**cocho\fR	cochlear coordinates (in millimeters)
\f(CR**deg\fR	key-related relative scale degree
\f(CR**degree\fR	key-related absolute scale degree
\f(CR**freq\fR	fundamental frequency (in hertz)
\f(CR**fret\fR	fretted-instrument pitch tablature
\f(CR**kern\fR	core pitch/duration representation
\f(CR**MIDI\fR	Music Instrument Digital Interface tablature
\f(CR**pc\fR	pitch class
\f(CR**pitch\fR	American National Standards Institute pitch notation
	  (e.g. \(odA#4\(cd)
\f(CR**semits\fR	equal-tempered semitones with respect to middle C=0
	  (e.g. 12 = C5)
\f(CR**solfa\fR	tonic solfa syllables
\f(CR**solfg\fR	French solf\o'e\(ga'ge system (fixed `doh')
\f(CR**specC\fR	spectral centroid (in hertz)
\f(CR**Tonh\fR	German pitch system
.TE
.sp 1
.ce
.I "Input representations processed by \fBsynco\fR."
.sp 1
.sp 1
The resulting \f(CR**synco\fR
spine contains numerical values, where zero represents no metric syncopation
and successively increasing values represent increasing amounts of
metric syncopation.
.sp 1
.sp 1
The
.B "synco"
command implements a definition of metric syncopation inspired by the
work of Lee and Longuet-Higgins (see REFERENCES).
In brief, metric syncopation may be defined as a moment where an expected
metric stress is absent.
More specifically, a metrically syncopated moment is defined as occurring
when no note-onset happens at a moment whose metric position is more
important than that of the most recent note onset.
For example, where a note onset occurs on the second beat of a quadruple meter,
and is not followed by a note onset on the third beat, the third beat is
deemed metrically syncopated because it occupies a higher metric position than
the previous onset.
.sp 1
.sp 1
The numerical output values generated by
.B "synco"
are calculated as the logarithm
of the metric position of the previous onset minus the logarithm
of the metric position of the current moment -- where the
current moment has no note onset, and coincides with a higher
metric position than the previous onset.
The use of the logarithmic difference weights the output values
so that missing downbeats at the beginning of a measure produce a greater
metric syncopation value than lesser beats or sub-beats in the measure.
In addition, metric syncopation is greater when the difference in
metric position between the previous onset and the current moment
is greatest.
(See EXAMPLES.)
.sp 1
.sp 1
If more than one musical part is given as input,
.B "synco"
responds to the
.I "aggregate"
rhythmic structure -- as though all of the parts were amalgamated
into a single rhythmic stream.
By itself, a single musical part may evoke considerable metric syncopation,
but in combination with other parts, metrically syncopated
moments are typically fewer.
In short, running
.B "synco"
on a multi-part score normally produces different results from running
.B "synco"
on each part individually.
.sp 1
.sp 1
.B "synco"
monitors the input in order to determine the Humdrum
.I "time-base"
-- if encoded.
Specifically,
.B "synco"
checks to ensure that the time-base is not excessively short.
It is possible to have time-base values that exceed the temporal resolving
power of human listeners.
For example, if an onset appears a thirty-second duration
prior to an expected down-beat, listeners are apt to hear the
displaced onset as occurring
.B "on"
the beat rather than being a very short syncopation.
\(odExcessively short\(cd is operationally defined as a time-base
resolution shorter than a sixteenth note.
In such cases,
.B "synco"
issues a warning, noting that the time-base may be too short.
.sp 1
.sp 1
Note that only one metrically syncopated moment can happen following a
given note onset;
subsequent syncopated moments require the intervention of another note onset.
By way of example, a note occurring immediately prior to an absent
major downbeat, will not also cause syncopated moments to arise
for other beats within a measure containing only rests.
In short, two metrically syncopated moments can't occur without some note
onset intervening.
.sp 1
.sp 1
It is recommended that output files produced using the
.B "synco"
command should be given names with the distinguishing `.syn' extension.
.in -2
.sp 1
.sp 1
.SH "OPTIONS"
.in +2
The
.B "synco"
command provides the following options:
.sp 1
.TS
l l.
\fB-e\fR	echo the input in the output
\fB-h\fR	displays a help screen summarizing the command syntax
.TE
.sp 1
Options are specified in the command line.
.sp 1
.sp 1
If the
.B "-e"
option is invoked, the output will echo all of the input spines along
with the \f(CR**synco\fR output.
.in -2
.sp 1
.sp 1
.SH "EXAMPLES"
.in +2
The following two examples illustrate the use of
.B "synco."
In both examples, the left-most spines represent the input,
and the right-most spine (labelled \f(CR**synco\fR) represents
the corresponding output.
.sp 1
.sp 1
The first example shows the minimum input of a single \f(CR**pitch\fR
input plus the metric position information (\f(CR**metpos\fR).
The meter signature is 2/4 and the time-base is an eighth duration.
(See the
.B "timebase (1)"
command.)
Hence there are 4 data records per measure.
The first beat in each measure is assigned the metric position "1";
the second beat is assigned the metric position "2";
and the second half of each beat is assigned the metric position "3".
Zero values in the \f(CR**synco\fR spine indicate the absence of any
syncopation.
In measure 3, a single syncopated moment happens at beat 2.
The output was produced using the simple command: \f(CRsynco inputfile\fR.
.in +2
.sp 1
.TS
l s s
l l l.
!! Example #1
**pitch	**metpos	**synco
*M2/4	*M2/4	*
*tb8	*tb8	*tb8
\(eq1	\(eq1	\(eq1
r	1	0
\.	3	0
r	2	0
A4	3	0
\(eq2	\(eq2	\(eq2
G4	1	0
\.	3	0
B4	2	0
r	3	0
\(eq3	\(eq3	\(eq3
C5	1	0
C5	3	0
\.	2	0.41
B4	3	0
\(eq4	\(eq4	\(eq4
*-	*-	*-
.TE
.sp 1
.in -2
In the following example, two metrically syncopated moments are
evident.
Notice that the rhythmic information for the two \f(CR**kern\fR
spines is amalgamated, and that the non-pitch spine (\f(CR**foo\fR)
has no affect on the processing.
.in +2
.sp 1
.TS
l s s l l
l l l l l.
!! Example #2
**foo	**kern	**metpos	**kern	**synco
*	*M2/4	*M2/4	*M2/4	*
*	*tb8	*tb8	*tb8	*tb8
A	=1	=1	=1	=1
A	4r	1	4r	0
A	.	4	.	0
A	.	3	.	0
A	.	4	.	0
A	8r	2	8r	0
A	.	4	.	0
A	[8a	3	[8a	0
A	.	4	.	0
A	=2	=2	=2	=2
A	4a]	1	8a]	1.10
A	.	4	.	0
A	.	3	8a	0
A	.	4	.	0
A	8b	2	8r	0
A	.	4	.	0
A	8r	3	8b	0
A	.	4	.	0
A	=3	=3	=3	=3
A	8cc	1	8cc	0
A	.	4	.	0
A	4.cc	3	4cc	0
A	.	4	.	0
A	.	2	.	0.41
A	.	4	.	0
A	.	3	8b	0
A	.	4	.	0
A	=4	=4	=4	=4
*-	*-	*-	*-	*-
.TE
.sp 1
.in -2
.in -2
.SH "PORTABILITY"
.in +2
\s-1DOS\s+1 2.0 and up, with the \s-1MKS\s+1 Toolkit.
\s-1OS/2\s+1 with the \s-1MKS\s+1 Toolkit.
\s-1UNIX\s+1 systems supporting the
.I "Korn"
shell or
.I "Bourne"
shell command interpreters, and revised
.I "awk"
(1985).
.in -2
.sp 1
.sp 1
.SH "SEE ALSO"
.in +2
\fBmetpos\fR (1), \fB**synco\fR (2), \fBtimebase\fR (1), \fBurrhythm\fR (1)
.in -2
.sp 1
.sp 1
.SH "REFERENCES"
.in +2
Longuet-Higgins, H. C., & Lee, C. S.
\(odThe perception of musical rhythms,\(cd
.I "Perception,"
Vol. 11 (1982) pp. 115-128.
.in -2


toolkit/awk/synco.awk
#####################################################################
##					SYNCO.AWK
# Programmed by: David Huron         Date: ?
# Copyright (c) 1994 David Huron
#
# Modifications:
#    Date:   Programmer:         	Description:
# Aug 2/94	Tim Racinsky	Modified parse_command to work with getopts
#
# Main Functions used:
#	Parse_command()		Store_indicators()		Store_new_interps()
#	Process_indicators()	Ins_array_pos()		Del_array_pos()
#	Exchange_spines()		Initialize()
#
#
#	VARIABLES:
#
#	options			holds options passed to this program
#	output_timebase	flag to indicate timebase interpretation should be
#					printed
#	note_onset		indicates whether the current sonority has a note onset
#	first_onset		indicates whether an onset has happened yet
#					(syncopated moments can happen only after an onset)
#	no_interpretation_yet	indicates whether first exclusive interpretat
#					has been encountered
#	current_no_of_spines	current number of spines in the input
#	metpos_spine		the input field number for **metpos information
#	onset_metpos_value	holds the current metric position value from **metpos
#					input
#	timebase			holds the current time-base value (*tb...)
#	processing		true when the input stream is suitable for calculating
#					and printing an output
#	ARRAYS:
#
#	target_interp[]	exclusive interpretations to be processed
#	process_spine[]	which data columns should be processed
#	path_indicator[]	contains successive spine-path record tokens
#	arraya[]			temporary array
#	spine_path_record	true when the current record is a spine-path record
#
BEGIN {
	#
	# Flags, variables, and constants:
	#
	FS = OFS = "\t"		# Set field separator to tab.
	YES = TRUE = 1;   NO = FALSE = 0
	#
	# Pipe all errors to this variable so that error messages are printed
	# to the terminal always
	#
	"echo $CON" | getline stderr
	close("echo $CON")
	options = ""
	cont_tie_reg = "\\]|_"
	#
	initialize()
	#
	target_interp[1]  = "**pitch"
	target_interp[2]  = "**barks"
	target_interp[3]  = "**cbr"
	target_interp[4]  = "**cents"
	target_interp[5]  = "**cocho"
	target_interp[6]  = "**deg"
	target_interp[7]  = "**degree"
	target_interp[8]  = "**freq"
	target_interp[9]  = "**fret"
	target_interp[10] = "**kern"
	target_interp[11] = "**mels"
	target_interp[12] = "**pc"
	target_interp[13] = "**Tonh"
	target_interp[14] = "**pu"
	target_interp[15] = "**semits"
	target_interp[16] = "**solfa"
	target_interp[17] = "**solfg"
	target_interp[18] = "**specC"
	#
	# Determine the options specified by the user, passed from the shell script
	#
	parse_command()
	}
############################################################################
#				              MAIN
############################################################################

{
if (FNR == 1)
	{
	#
	# In the case of multiple input files, the following variables are reset.
	#
	initialize()
	#
	}
if ($0 ~ /^!!/)
	{
	# Echo global comments
	#
	print $0
	next
	#
	}
else if ($0 ~ /^!([^!]|$)/)
	{
	# Echo local comments as a null local comment.
	#
	if (options ~ /e/) print $0 "\t!"
	else print "!"
	next
	#
	}
else if ($0 ~ /^\*/)
	{
	#
	# Process the first interpretation record.  (By definition, the
	# first interpretation record must be an exclusive interpretation.)
	#
	if (no_interpretation_yet)
		{
		current_no_of_spines = NF
		no_interpretation_yet = FALSE
		store_new_interps()
		}
	#
	# Otherwise, check to see if it is a spine-path record
	#
	else
		{
		spine_path_record = FALSE
		for (i=1; i<=current_no_of_spines ; i++)
			{			
			if ($i ~ /^\*\+$|^\*\-$|^\*\^$|^\*v$|^\*x$/)
				{
				spine_path_record = TRUE 
				break
				}
			}
		#
		# If it is a spine-path record, store and process the indicators
		#
		if (spine_path_record)
			{
			store_indicators()
			process_indicators()
			#
			# Reset any necessary variables if all spines terminate
			#
			if (current_no_of_spines == 0)
				{
				no_interpretation_yet = TRUE
				}
			}
		#
		# Otherwise, process the interpretations (both tandem & exclusive)
		#
		else
			{
			store_new_interps()
			}		
		}
	}
else
	{
	########################################################################
	#
	#	Process data records.
	#
	########################################################################
	#
	# Don't process data unless you have to.
	#
	if (!processing) next
	#
	# Determine whether the current sonority has a note onset.
	#
	note_onset = FALSE
	for (i=1; i<=current_no_of_spines; i++)
		{
		if (!process_spine[i]) continue
		# Echo barlines.
		if ($i ~ /^\=/)
			{
			if (options ~ /e/) print $0 "\t" $i
			else print $i
			next
			}
		#
		if ($i != "." && $i !~ /r/ && $i !~ cont_tie_reg)
			{
			note_onset = TRUE
			first_onset = TRUE
			# Record the associated metpos value.
			onset_metpos_value = $metpos_spine
			}
		}
	# Syncopations arise when no note-onset happens at a moment whose metric
	# position is more important than that of the most recent note onset.
	#
	if (!note_onset && first_onset && $metpos_spine < onset_metpos_value)
		{
		# Avoid taking the logarithm of zero.
		#
		if (onset_metpos_value != 0 && $metpos_spine != 0)
			{
			if (options ~ /e/) printf("%s\t%4.2f\n",$0,\
					log(onset_metpos_value) - log($metpos_spine))
			else printf("%4.2f\n",\
					log(onset_metpos_value) - log($metpos_spine))
			}
		else
			{
			print "synco: ERROR: Illegal metric position value `0' found" \
							" near line: " NR > stderr
			exit
			}
		#
		# Syncopated moments can happen only after an onset;
		# subsequent syncopated moments will require another onset.
		# (i.e. two syncopated moments can't occur without some note
		#  onset intervening).
		first_onset = FALSE
		}
	else
		{
		if (options ~ /e/) print $0 "\t0"
		else print "0"
		}
	}
}

############################################################################
#				FUNCTIONS USED IN SYNCO.AWK
############################################################################

############################################################################
##				Function Parse_command
# This function inspects the array ARGV in order to determine the correct
# options that the user has specified.
#
function parse_command()
	{
	#
	# Process any necessary variables
	#
	if (ARGV[2] != "null") options = ARGV[2]
	ARGV[1] = ARGV[2] = ""
	}

################################################################
##			Function Store_new_interps
#	This function stores the new interpretations found in an
# interpretation record for each spine.
#
function store_new_interps(      j,i)
	{
	#
	# Process exclusive interpretations
	#
	if ($0 ~ /^\*\*/ || $0 ~ /\t\*\*/)
		{
		# Process data only if a **metpos spine and at least
		# one **[target_interp] spine are found in the input.
		#
		processing = FALSE
		#
		# Identify whether an active **metpos spine has been changed
		# to some other interpretation.
		#
		#if (metpos_spine == j && $j !~ /^\*\*metpos$/) metpos_spine = 0
		#
		# Identify any **metpos spine.
		#
		for (j = 1; j <= current_no_of_spines; j++)
			{
			if ($j ~ /^\*\*metpos$/) metpos_spine = j
			}
		#
		# Identify all input spines that contain target interpretations.
		#
		for (j = 1; j <= current_no_of_spines; j++)
			{
			if ($j ~ /^\*\*/) process_spine[j] = FALSE
			#
			for (i in target_interp)
				{
				if ($j == target_interp[i]) process_spine[j] = YES
				if (process_spine[j] && metpos_spine !=0) processing = YES
				}
			}
		if (processing)
			{
			if (options ~ /e/) print $0 "\t**synco"
			else print "**synco"
			}
		}
	#
	# Process tandem interpretations.
	#
	# If a timebase tandem interpretation is present, check to ensure
	# that the timebase is not excessively short.  It is possible to
	# have timebase values that exceed the temporal resolving power of
	# listeners.  For example, if an onset appears a thirty-second
	# duration prior to an expected down-beat, the listener is apt to
	# hear the displaced onset as occuring ON the beat rather than being
	# a very short syncopation.  "Excessively short" is defined as a
	# timebase resolution shorter than a 16th note.
	#
	else
		{
		if (!processing) return
		output_timebase = FALSE
		for (j = 1; j <= current_no_of_spines; j++)
			{
			if ($j ~ /^\*tb[1-9]?[0-9]*$/)
				{
				timebase = $j
				output_timebase = YES
				gsub("*tb","",timebase)
				if (timebase+0 > 16) print "synco: WARNING: Time-base very"\
				" short; listener's temporal resolution exceeded?" > stderr
				}
			}
		if (output_timebase)
			{
			if (options ~ /e/) print $0 "\t*tb" timebase
			else print "*tb" timebase
			}
		else # Echo all other tandem interpretations.
			{
			if (options ~ /e/) print $0 "\t*"
			else print "*"
			}
		}
	}

##############################################################
##			Function Store_indicators
#	This function allows the spine-path indicators for the 
# current record to be stored in the array 'path_indicator' so
# that they may be used later.
#
function store_indicators(   i)
	{
	for (i = 1; i <= current_no_of_spines; i++)
		{
		#
		# If the current field is a null interpretation
		#
		if ($i == "*") path_indicator[i] = "*"
		#
		# If the current field is a spine-path indicator, store it
		#
		else path_indicator[i] = substr($i,2,1)
		}
	} 

##################################################################
##				Function Process_indicators
#	This function takes the spine-path indicators that were stored
# in the array 'path_indicator' in the function 'store_indicators'
# and manipulates the arrays 'path_indicator' and 'process_spine'
# according to the contents of the array 'path_indicator'.
#
function process_indicators(    i,k,end)
	{
	# If all spine-path indicators are spine terminators, print
	# the appropriate output record, and get the next input record.
	#
	end = TRUE
	for (i=1; i<=current_no_of_spines; i++) if ($i != "*-") end = FALSE
	if (processing)
		{
		if (end)
			{
			if (options ~/e/) print $0 "\t*-"
			else print "*-"
			initialize()
			return
			}
		# Otherwise print the appropriate spine-path record.
		#
		else
			{
			if (options ~ /e/) print $0 "\t*"
			else print "*"
			}
		}
	#
	i = 1
	#
	# Loop through the entire array of 'path_indicator'
	# NOTE: This array may be altered during the process
	#
	while (i <= current_no_of_spines)
		{
		#
		# If a spine-split indicator occurred
		#
		if (path_indicator[i] == "^")
			{
			#
			# Insert a new array position in 'path_indicator'
			# and 'process_spine'
			#
			ins_array_pos(i+1)
			path_indicator[i+1] = "*"
			process_spine[i+1] = process_spine[i]	# Copy interpretations
											# to the new spine.
			if (metpos_spine > i) metpos_spine++
			}
		#
		# If a spine-termination indicator occurred
		#
		if (path_indicator[i] == "-")
			{
			#
			# Delete that spine from the arrays and look at the same
			# position the next time through the loop
			#
			del_array_pos(i)
			if (metpos_spine > i) metpos_spine--
			i--
			}
		#
		# If a spine-exchange indicator occurred
		#
		if (path_indicator[i] == "x")
			{
			#
			# Exchange the paths
			#
			exchange_spines()
			}
		#
		# If a join-path indicator occurred
		#
		if (path_indicator[i] == "v")
			{
			del_array_pos(i+1)	# Delete NEXT spine (must be join-path
							# spine according to humdrum).
			if (metpos_spine > i) metpos_spine--
			#
			# If the next indicator is not a join-path indicator
			# then return to original conditions
			#
			if (path_indicator[i+1] != "v")
				{
				path_indicator[i] = "*"
				}
			#
			# Otherwise, look at current and next paths to join
			#
			else i--
			}
		#
		# If a new-path indicator occurred
		#
		if (path_indicator[i] == "+")
			{
			#
			# Insert a new position in the arrays
			#
			ins_array_pos(i+1)
			process_spine[i+1] = ""
			if (metpos_spine > i) metpos_spine++
			}
		i++
		}
	}

############################################################
##			Function Ins_array_pos
#	This function inserts a new position in the arrays 'path_indicator'
# and 'process_spine' and copies elements so that order is preserved
#
function ins_array_pos(current_element,     j) 
	{
	for (j = current_no_of_spines+1; j > current_element; j--)
		{
		path_indicator[j] = path_indicator[j-1]
		process_spine[j] = process_spine[j-1]
		}
	path_indicator[current_element-1] = "*"	# (Avoid possibility of
	path_indicator[current_element] = "*"	#  reprocessing the old spine.)
	current_no_of_spines++
	}

##############################################################
##			Function Del_array_pos
#	Performs the opposite of function 'ins_array_pos'.
#
function del_array_pos(current_element,     j)
	{
	for (j = current_element; j < current_no_of_spines; j++)
		{
		path_indicator[j] = path_indicator[j+1]
		process_spine[j] = process_spine[j+1]
		}
	delete path_indicator[j]
	delete process_spine[j]
	current_no_of_spines--
	} 

###################################################################
##				Function Exchange_spines
#	This function exhanges two spines by exchanging the corresponding
# elements in 'process_spine'
#
function exchange_spines(    j,count,arraya,temp)
	{
	count = 1
	# Determine which spines are exchanged
	for (j = 1; j <= current_no_of_spines; j += 1)
		{
		if (path_indicator[j] == "x")
			{
			arraya[count] = j
			count += 1
			path_indicator[j] = "*"
			}
		}
	# Exchange interpretations between the two spines
	temp = process_spine[arraya[1]]
	process_spine[arraya[1]] = process_spine[arraya[2]]
	process_spine[arraya[2]] = temp
	}
###################################################################
##				Function Initialize
#	This function initializes or resets various values for a new input.
function initialize()
	{
	note_onset = FALSE
	first_onset = FALSE
	no_interpretation_yet = TRUE
	current_no_of_spines = 0
	metpos_spine = 0
	for (i in process_spine) process_spine[i] = FALSE
	}
